Fix $bulk-delete bugs with custom search parameter

tarunmathew12 opened the pull request:
## Description
The PR will address an issue of the search parameter status going out of sync after $bulk-delete removes search parameter resources. 

## Related issues
Addresses [issue #121676].
[Bug 121676](https://microsofthealth.visualstudio.com/Health/_workitems/edit/121676): Hard delete custom search parameter not cleaning up all the data

## Testing
Tested manually and also through adding E2E test.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- When changing or adding behavior, if your code modifies the system design or changes design assumptions, please create and include an [ADR](https://github.com/microsoft/fhir-server/blob/main/docs/arch).
- [ ] CI is green before merge [![Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_apis/build/status/CI%20Build%20%26%20Deploy?branchName=main)](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/main/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/main/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

tarunmathew12 commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BulkDeleteTests.cs:
```diff
@@ -357,6 +357,61 @@
+                        {
+                            await _fhirClient.DeleteAsync(entry.Resource);
+                        }
+                        catch
+                        {
+                            // Ignore and let the cleanup complete.
+                        }
```
Need to catch everything to clean up resources.

tarunmathew12 commented:
/azp run

tarunmathew12 commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BulkDeleteTests.cs:
```diff
@@ -357,6 +362,111 @@
+                            {
+                                await _fhirClient.HardDeleteAsync(resourceToDelete);
+                            }
+                        }
```
Won't fix.

tarunmathew12 commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BulkDeleteTests.cs:
```diff
@@ -357,6 +360,151 @@
+                         }
+
+                         await Task.Delay(2000);
+                         foreach (var resource in resourcesCreated)
+                         {
+                             var response = await _fhirClient.SearchAsync(
+                                 ResourceType.SearchParameter,
+                                 $"url={((SearchParameter)resource).Url}");
+                             Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+                             var resourcesRetrieved = response.Resource?.Entry
+                                 .Select(x => x.Resource)
+                                 .Where(x => x.TypeName?.Equals(ResourceType.SearchParameter.ToString(), StringComparison.OrdinalIgnoreCase) ?? false)
+                                 .ToList();
+                             Assert.Single(resourcesRetrieved);
+                         }
```
Won't fix.

tarunmathew12 commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BulkDeleteTests.cs:
```diff
@@ -357,6 +364,277 @@
+                            DebugOutput(urls.ToString());
+                        }
+
+                        foreach (var resource in resources)
+                        {
+                            var url = ((SearchParameter)resource).Url;
+                            var response = await _fhirClient.SearchAsync(
+                                ResourceType.SearchParameter,
+                                $"url={url}");
+                            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+                            var resourceToDelete = response.Resource?.Entry?
+                                .Select(x => x.Resource)
+                                .Where(x => x.TypeName?.Equals(ResourceType.SearchParameter.ToString(), StringComparison.OrdinalIgnoreCase) ?? false)
+                                .FirstOrDefault();
+                            if (resourceToDelete != null)
+                            {
+                                try
+                                {
+                                    DebugOutput($"Url deleting: {url}");
+                                    await _fhirClient.HardDeleteAsync(resourceToDelete);
+                                    DebugOutput($"Url deleted: {url}");
+                                }
+                                catch (Exception ex)
+                                {
+                                    DebugOutput($"Url delete failed: {url}{Environment.NewLine}{ex}");
+                                    throw;
+                                }
+                            }
+                            else
+                            {
+                                DebugOutput($"Url not found: {url}");
+                            }
+                        }
```
Won't fix.

tarunmathew12 commented on src/Microsoft.Health.Fhir.Core/Features/Operations/BulkDelete/BulkDeleteProcessingJob.cs:
```diff
@@ -132,5 +148,40 @@
+                        {
+                            return _modelInfoProvider.ToTypedElement(x.RawResource).GetStringScalar("url");
+                        }
+                        catch (Exception ex)
+                        {
+                            _logger.LogError(ex, $"Failed to extract the url from the resource, '{x.ResourceId}'.");
+                            return null;
+                        }
```
Won't fix. We need to catch all exceptions and log the exception regardless of the exception type.

tarunmathew12 commented on src/Microsoft.Health.Fhir.Core/Features/Search/Parameters/SearchParameterOperations.cs:
```diff
@@ -316,5 +351,51 @@
+                                    SearchParameterStatus.PendingDelete);
+                            });
+                    }
+                    catch (Exception ex)
+                    {
+                        // Note: ignore the exception and continue updating the rest.
+                        _logger.LogError(ex, $"Failed to update the status of search parameters:{Environment.NewLine}{urlsString}");
+                    }
```
Won't fix. We need to catch all exceptions and log the exception regardless of the exception type.

tarunmathew12 commented on test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/BulkDeleteTests.cs:
```diff
@@ -356,6 +364,328 @@
+                await CleanupAsync(resourcesToCreate);
+
+                // Create search parameter resources.
+                var resources = await CreateAsync(resourcesToCreate);
```
Fixed as suggested

LTA-Thinking commented on src/Microsoft.Health.Fhir.Core/Features/Definition/SearchParameterDefinitionManager.cs:
```diff
@@ -271,10 +271,12 @@ public void DeleteSearchParameter(string url, bool calculateHash = true)
+                if (TypeLookup[resourceType].TryRemove(searchParameterInfo.Code, out var removedParam))
                 {
-                    _logger.LogError("Error, Search Param {RemovedParam} removed from Search Param Definition manager.  It does not match deleted Search Param {Url}", removedParam.Url, url);
+                    if (removedParam?.Url != searchParameterInfo.Url)
```
I agree with the code scanner, the two if statements should be combined.

tarunmathew12 replied:
Fixed as suggested.

LTA-Thinking reviewed (changes requested):
I don't believe the response from the deletion service can be a list of resource wrappers due to memory limitations.

tarunmathew12 commented:
> I don't believe the response from the deletion service can be a list of resource wrappers due to memory limitations.

We have reverted it back to returning the stats dictionary. We have also added an optional parameter to DeleteMultipleAsync to store deleted search parameters.

tarunmathew12 commented:
/azp run

tarunmathew12 commented:
/azp run

tarunmathew12 commented:
/azp run

LTA-Thinking commented on src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Delete/DeletionService.cs:
```diff
@@ -146,7 +143,7 @@ private async Task<IDictionary<string, long>> DeleteMultipleAsyncInternal(Condit
                     continuationToken,
                     versionType: request.VersionType,
-                    onlyIds: true,
+                    onlyIds: !string.Equals(request.ResourceType, KnownResourceTypes.SearchParameter, StringComparison.OrdinalIgnoreCase),
```
This is also needed on line 239

LTA-Thinking commented:
Has conditional delete been tested to see how it handles deleting search parameters?

LTA-Thinking commented on src/Microsoft.Health.Fhir.Core/Features/Operations/BulkDelete/BulkDeleteProcessingJob.cs:
```diff
@@ -105,7 +122,20 @@ public async Task<string> ExecuteAsync(JobInfo jobInfo, CancellationToken cancel
+                        Content = CreateNotificationContent(deletedSearchParameters),
+                    };
+
+                    await _mediator.Publish(notification, cancellationToken);
```
My only concern with this method of updating the search parameters is that it isn't safe if the service shuts down. If the notification isn't fully handled the DB can be left in a bad state where the Search Parameter resources are deleted, but the Search Parameter table hasn't been updated yet.
